truffle/src/main/ruby/core/thread.rb

```diff
@@ -352,15 +352,38 @@ def freeze
 end
 
 class ThreadGroup
+  def initialize
+    @enclosed = false
+  end
 
-  attr_reader :list
+  def enclose
+    @enclosed = true
+  end
 
-  def initialize
-    @list = []
+  def enclosed?
+    @enclosed
   end
 
   def add(thread)
-    @list.push thread
+    raise ThreadError, "can't move to the frozen thread group" if self.frozen?
+    raise ThreadError, "can't move to the enclosed thread group" if self.enclosed?
+
+    from_tg = thread.group
+    return nil unless from_tg
+    raise ThreadError, "can't move from the frozen thread group" if from_tg.frozen?
+    raise ThreadError, "can't move from the enclosed thread group" if from_tg.enclosed?
+
+    Truffle.invoke_primitive :thread_set_group, thread, self
+    self
   end
 
+  def list
+    Thread.list.find_all { |th| th.group == self }
```

Comment:
@eregon I made this a calculated field only to match MRI. It might have some advantages: harder to change a calculated field versus a instance var, etc.